### ISSUE TITLE
[MNG-3655] Allow multiple local repositories

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#5373]
**[Ittay Dror](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ittayd)** opened **[MNG-3655](https://issues.apache.org/jira/browse/MNG-3655?redirect=false)** and commented

In some environments, branches are rarely used. This means that if a developer wishes to work in parallel on two features, he checks out HEAD into two different locations. The problem is that using 'mvn install' in one checkout will overwrite the result of 'mvn install' in another. Of course one can write poms so that the version contains some classifier and then use 'mvn -Dartifact-classifier=first-checkout install', or, read from a file. Both are tedious.

Instead, it would be good to be able to tell maven to first consider some path under the checkout before trying a global local repository (for external artifacts).

To make this work when running mvn from a module subdir, maybe allow to write settings.xml in the root directory of the checkout. Then, maven should climb the directory structure until locating settings.xml (or reaching the global root directory) and read there. Using settings.xml in such a way has other benefits that it can be under version control. settings.xml will then be able to specify a list of local repositories, some absolute paths, some relative to it.

Another approach could be to allow this list of local repositories in the global settings.xml file and have an entry in each module's pom indicating where it is relative to the local repository (like the parent path attribute)

---

**Issue Links:**
- [MNG-4709](https://issues.apache.org/jira/browse/MNG-4709) Separate local repository for each remote repository
 (_**"is duplicated by"**_)
- [MNG-4071](https://issues.apache.org/jira/browse/MNG-4071) Support additional read-only local repositories
 (_**"is duplicated by"**_)
- [MNG-6270](https://issues.apache.org/jira/browse/MNG-6270) Store snapshots in a separate local repository

- [MINSTALL-126](https://issues.apache.org/jira/browse/MINSTALL-126) localRepositoryPath for install like for install-file

- [MRESOLVER-253](https://issues.apache.org/jira/browse/MRESOLVER-253) Dynamic prefixes for LRM
 (_**"depends upon"**_)
- [MNG-7499](https://issues.apache.org/jira/browse/MNG-7499) Upgrade Maven Resolver to 1.8.1
 (_**"depends upon"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=27823062)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1)

15 votes, 25 watchers

**Comment by jira-importer** [comment]
**[Wendy Smoak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wsmoak)** commented

As a workaround, you can use -Dmaven.repo.local=/path/to/repo on the command line to point at an alternate local repo.

**Comment by jira-importer** [comment]
**[Ittay Dror](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ittayd)** commented

I know. The problem with such solutions is that it creates a usability issue (always need to pass the variable) and also all maven plugins and external artifacts will be downloaded again, taking up space and time (think of checking out to do a simple bug fix)

**Comment by jira-importer** [comment]
**[Ittay Dror](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ittayd)** commented

This is not a duplicate of 2802. I don't need several local repositories to support concurrently running maven, I need it to create a hierarchy of repositories where artifacts are located.

Imagine a simple project, creating 2 jars named foo-bar-1.0.jar and foo-zoo-1.0.jar. zoo depends on bar.

I checkout the projects to a directory named named 'foo' (subdirectories 'bar' and 'zoo')

On Monday I come to work, and start developing a new feature. Happily  coding and building. I need to install to local repository since 'zoo' depends on 'bar. All is well.

On Tuesday, my boss comes and tells me there's an urgent bug fix (be it in some client, or simply because QA wants to try out the current iteration of the projects but are blocked).

What do I do? I checkout the projects to a directory 'foo_bugfix' and start correcting the bug.

Now, If I build, the artifacts from 'foo_bugfix' will override those from 'foo'.

Maybe this doesn't sound like much of a problem for this toy example, when I return to work in 'foo', just do a complete build.

But, it becomes a headache for large projects with tens of modules. In such cases, a full rebuild may take a long time. Also, working with branches becomes an issue, since on every checkout from a branch, I need to change version numbers to avoid conflicts.

What I'm suggesting is to allow a user like me to define a local repository that is relative to the current build directory. All problems described above are gone. However, doing just that means that on every checkout I have to wait for all artifacts (plugins, 3rd party, nightly build) to be downloaded. So instead I want the best of both worlds: allow me to define a list of repositories that are examined for artifacts, in order. When installing an artifact, write it to the first repository on the list. Then, I can first setup a shared local repository and afterwards add a reference to a local repository with relative path where further downloaded/installed artifacts are put.

Another approach is to have two definitions: one for 'local repository' and the other for 'project repository'. All items downloaded go to 'local repository' and all installed items go to 'project repository', maybe with a way of configuring maven so that some downloaded items, from a nightly build, go to 'project repository'.

Btw, about nightly builds, it will be good if remote repositories could have a mechanism where if I checkout a branch, maven first tries to download from a repository specifically for that branch (maybe the repository url can have some configuration value, but as long as it's not something that I need to update on every checkout)

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

apologies, I linked the wrong issue. What I thought I was linking to was this:
https://cwiki.apache.org/confluence/display/MAVENOLD/Local+repository+separation

Would that address your needs?

**Comment by jira-importer** [comment]
**[Ittay Dror](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ittayd)** commented

sounds good

**Comment by jira-importer** [comment]
**[Mark](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cinquero)** commented

Yes, that would be great. Would improve repeatability of builds and eliminate re-downloading of artifacts where repeatability is an issue and there is no artifact cache/proxy/daemon.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Use a local Nexus instance to avoid downloading over and over again.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Why can't you just use a separate settings file and `-Dmaven.repo.local`?

**Comment by jira-importer** [comment]
**[Dimo Velev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dimovelev)** commented

I am interested in splitting local cache of releases and snapshots. This way, I could have one common releases local cache (on a jenkins node, or maybe in a kubernetes volume etc) and multiple snapshots local caches for each build. This would greatly improve the isolation while building and reduce the duplication of released artifacts.

Really hope that this makes it into maven at some point.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I believe that this issue is fully covered by MRESOLVER-253.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas), this is now possible with Resolver 1.8.1. Can you complete this issue?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

As documented, this is now supported from Resolver 1.8.1 ("branched development"), that will be present in Maven 3.9.0+

See https://maven.apache.org/resolver/local-repository.html#use-cases
